ibm_cloud_sdk_core/token_manager.py

```diff
@@ -0,0 +1,224 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from abc import ABC, abstractmethod
+from threading import Lock
+
+import time
+import requests
+
+from .api_exception import ApiException
+
+
+class TokenManager(ABC):
+    """An abstract class to contain functionality for parsing, storing, and requesting tokens.
+
+    get_token will retrieve a new token from the url in case the that there is no existing token,
+    or the previous token has expired. Child classes will implement request_token, which will do
+    the actual acquisition of a new token.
+
+    Args:
+        url: The url to request tokens from.
+
+    Keyword Args:
+        disable_ssl_verification: A flag that indicates whether verification of
+            the server's SSL certificate should be disabled or not. Defaults to False.
+
+    Attributes:
+        url (str): The url to request tokens from.
+        disable_ssl_verification (bool): A flag that indicates whether verification of
+        the server's SSL certificate should be disabled or not.
+        expire_time (int): The time in epoch seconds when the current stored token will expire.
+        refresh_time (int): The time in epoch seconds when the current stored token should be refreshed.
+        request_time (int): The time the last outstanding token request was issued
+        lock (Lock): Lock variable to serialize access to refresh/request times
+        http_config (dict): A dictionary containing values that control the timeout, proxies, and etc of HTTP requests.
+    """
+
+    def __init__(
+            self,
+            url: str,
+            *,
+            disable_ssl_verification: bool = False
+    ):
+        self.url = url
+        self.disable_ssl_verification = disable_ssl_verification
+        self.expire_time = 0
+        self.refresh_time = 0
+        self.request_time = 0
+        self.lock = Lock()
+        self.http_config = {}
+
+    def get_token(self) -> str:
+        """Get a token to be used for authentication.
+
+        The source of the token is determined by the following logic:
+        1.  a) If the current token is expired (or never fetched), make a request for one
+            b) If the current token should be refreshed, issue a refresh request
+        2. After any requests initiated above complete, return the stored token
+
+        Returns:
+            str: A valid access token
+        """
+        if self._is_token_expired():
+            self.paced_request_token()
+
+        if self._token_needs_refresh():
+            token_response = self.request_token()
+            self._save_token_info(token_response)
+
+        return self.extract_token_from_stored_response()
```

Comment:
I suppose this works but I think I would prefer that the `access_token` be defined as a property that is set by `_save_token_info` -- along with `expire_time` and `refresh_time` -- so that here we can simply have:
```
    return self.access_token
```

Reply:
Done in dc4162a

Mind that, this alters semantics slightly: https://github.com/IBM/python-sdk-core/blob/master/ibm_cloud_sdk_core/jwt_token_manager.py#L85 so a minor test change was necessary: https://github.com/IBM/python-sdk-core/pull/72/commits/dc4162adc3fec8d26d91fa0b6addb8b5f3c239e5#diff-6bc545e548b4e369e6823510e07c95ceR65